<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>8&nbsp; Nested random effects</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/09-crossed-random-effects.html" rel="next">
<link href="../materials/07-checking-assumptions.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-bce20943a2e3f93eb31aa65a0aa0ff1d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Mixed effects models</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/cambiotraining/stats-mixed-effects-models"> <i class="bi bi-github" role="img" aria-label="Bioinformatics Training Facility GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/bioinfocambs"> <i class="bi bi-twitter" role="img" aria-label="Bioinformatics Training Facility Twitter">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://genomic.social/@BioInfoCambs"> <i class="bi bi-mastodon" role="img" aria-label="Bioinformatics Training Facility Mastodon">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../materials/08-nested-random-effects.html">More complex designs</a></li><li class="breadcrumb-item"><a href="../materials/08-nested-random-effects.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Nested random effects</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Welcome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Overview</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data &amp; Setup</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Random effects</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/03-independence-psuedoreplication.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Independence &amp; pseudoreplication</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/04-random-effects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introducing random effects</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/05-fitting-mixed-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Fitting mixed models</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Interpreting mixed models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/06-significance-and-model-comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Significance &amp; model comparison</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/07-checking-assumptions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Checking assumptions</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">More complex designs</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/08-nested-random-effects.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Nested random effects</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/09-crossed-random-effects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Crossed random effects</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Bonus materials</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/10-generalised-mixed-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Generalised mixed models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/11-correlations-between-random-effects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Correlations between random effects</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/12-simulating-hierarchical-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Simulating hierarchical data</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#what-is-a-nested-random-effect" id="toc-what-is-a-nested-random-effect" class="nav-link active" data-scroll-target="#what-is-a-nested-random-effect"><span class="header-section-number">8.1</span> What is a nested random effect?</a></li>
  <li><a href="#fitting-a-three-level-model" id="toc-fitting-a-three-level-model" class="nav-link" data-scroll-target="#fitting-a-three-level-model"><span class="header-section-number">8.2</span> Fitting a three-level model</a>
  <ul class="collapse">
  <li><a href="#a-three-level-random-intercepts-model" id="toc-a-three-level-random-intercepts-model" class="nav-link" data-scroll-target="#a-three-level-random-intercepts-model"><span class="header-section-number">8.2.1</span> A three-level random intercepts model</a></li>
  <li><a href="#where-to-include-random-slopes" id="toc-where-to-include-random-slopes" class="nav-link" data-scroll-target="#where-to-include-random-slopes"><span class="header-section-number">8.2.2</span> Where to include random slopes?</a></li>
  </ul></li>
  <li><a href="#implicit-vs-explicit-nesting" id="toc-implicit-vs-explicit-nesting" class="nav-link" data-scroll-target="#implicit-vs-explicit-nesting"><span class="header-section-number">8.3</span> Implicit vs explicit nesting</a>
  <ul class="collapse">
  <li><a href="#the-pastes-dataset" id="toc-the-pastes-dataset" class="nav-link" data-scroll-target="#the-pastes-dataset"><span class="header-section-number">8.3.1</span> The Pastes dataset</a></li>
  <li><a href="#recoding-for-implicit-nesting" id="toc-recoding-for-implicit-nesting" class="nav-link" data-scroll-target="#recoding-for-implicit-nesting"><span class="header-section-number">8.3.2</span> Recoding for implicit nesting</a></li>
  <li><a href="#fitting-a-model-with-explicit-nesting" id="toc-fitting-a-model-with-explicit-nesting" class="nav-link" data-scroll-target="#fitting-a-model-with-explicit-nesting"><span class="header-section-number">8.3.3</span> Fitting a model with explicit nesting</a></li>
  <li><a href="#an-alternative-approach-the-ab-syntax" id="toc-an-alternative-approach-the-ab-syntax" class="nav-link" data-scroll-target="#an-alternative-approach-the-ab-syntax"><span class="header-section-number">8.3.4</span> An alternative approach: the A/B syntax</a></li>
  <li><a href="#which-option-is-best" id="toc-which-option-is-best" class="nav-link" data-scroll-target="#which-option-is-best"><span class="header-section-number">8.3.5</span> Which option is best?</a></li>
  </ul></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">8.4</span> Exercises</a>
  <ul class="collapse">
  <li><a href="#sec-exr_cake" id="toc-sec-exr_cake" class="nav-link" data-scroll-target="#sec-exr_cake"><span class="header-section-number">8.4.1</span> Cake</a></li>
  <li><a href="#sec-exr_parallel" id="toc-sec-exr_parallel" class="nav-link" data-scroll-target="#sec-exr_parallel"><span class="header-section-number">8.4.2</span> Parallel fibres</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">8.5</span> Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../materials/08-nested-random-effects.html">More complex designs</a></li><li class="breadcrumb-item"><a href="../materials/08-nested-random-effects.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Nested random effects</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Nested random effects</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Mixed effects models are also sometimes referred to as “hierarchical” or “multi-level” models. So far in these materials, we’ve only fitted two-level models, containing a single clustering variable or effect. Sometimes, however, there are random effects nested <em>inside</em> others.</p>
<section id="what-is-a-nested-random-effect" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="what-is-a-nested-random-effect"><span class="header-section-number">8.1</span> What is a nested random effect?</h2>
<p>Once we are at the stage of having multiple variables in our sample that create clusters or groups, it becomes relevant to consider the relationship that those clustering variables have to one another, to ensure that we’re fitting a model that properly represents our experimental design.</p>
<p>We describe factor B as being nested inside factor A, if each group/category of B only occurs within one group/category of factor A.</p>
<p>For instance, data on academic performance may be structured as children grouped within classrooms, with classrooms grouped within schools. A histology experiment might measure individual cells grouped within slices, with slices grouped within larger samples. Air pollution data might be measured at observation stations grouped within a particular city, with multiple cities per country.</p>
</section>
<section id="fitting-a-three-level-model" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="fitting-a-three-level-model"><span class="header-section-number">8.2</span> Fitting a three-level model</h2>
<p>Another classic example of nested random effects that would prompt a three-level model can be found in a clinical setting: within each hospital, there are multiple doctors, each of whom treats multiple patients. (Here, we will assume that each doctor only works at a single hospital, and that each patient is only treated by a single doctor.)</p>
<p>Here’s an image of how that experimental design looks. Level 1 is the patients, level 2 is the doctors, and level 3 is the hospitals.</p>
<p>This is, of course, a simplified version - we would hope that there are more than two hospitals, four doctors and eight patients in the full sample!</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/nested-patients1.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Experimental design</figcaption>
</figure>
</div>
<p>We have a single fixed predictor of interest <code>treatment</code> (for which there are two possible treatments, A or B), and some continuous response variable <code>outcome</code>.</p>
<p>What model would we fit to these data? Well, it gets a touch more complex now that we have multiple levels in this dataset.</p>
<section id="a-three-level-random-intercepts-model" class="level3" data-number="8.2.1">
<h3 data-number="8.2.1" class="anchored" data-anchor-id="a-three-level-random-intercepts-model"><span class="header-section-number">8.2.1</span> A three-level random intercepts model</h3>
<p>Let’s put random slopes to one side, since they take a bit more thought, and think about how we would fit just some random intercepts for now.</p>
<p>It would be appropriate to fit two sets of random intercepts in this model, one for each set of clusters we have. In this case, that means a set of intercepts for the doctors, and a set of intercepts for the hospital.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" aria-current="page">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>health <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/health.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>lme_health_intercepts <span class="ot">&lt;-</span> <span class="fu">lmer</span>(outcome <span class="sc">~</span> treatment <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>doctor) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>hospital),</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> health)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_health_intercepts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: outcome ~ treatment + (1 | doctor) + (1 | hospital)
   Data: health

REML criterion at convergence: 1848.5

Scaled residuals: 
     Min       1Q   Median       3Q      Max 
-2.56989 -0.66086  0.06162  0.67602  2.84690 

Random effects:
 Groups   Name        Variance Std.Dev.
 doctor   (Intercept)  2.1416  1.4634  
 hospital (Intercept)  0.1688  0.4108  
 Residual             26.3425  5.1325  
Number of obs: 300, groups:  doctor, 30; hospital, 5

Fixed effects:
                 Estimate Std. Error       df t value Pr(&gt;|t|)    
(Intercept)       26.6155     0.5299   8.4431   50.23 9.34e-12 ***
treatmentsurgery   6.2396     0.5926 269.0000   10.53  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Correlation of Fixed Effects:
            (Intr)
trtmntsrgry -0.559</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This produces a model with two random effects, namely, two sets of random intercepts.</p>
</section>
<section id="where-to-include-random-slopes" class="level3" data-number="8.2.2">
<h3 data-number="8.2.2" class="anchored" data-anchor-id="where-to-include-random-slopes"><span class="header-section-number">8.2.2</span> Where to include random slopes?</h3>
<p>Deciding what level(s) we want to fit random slopes at, requires us to think about what level of our hierarchy we’ve applied our <code>Treatment</code> variable at. We’ll get to that in a moment.</p>
<section id="predictor-varies-at-level-1" class="level4">
<h4 class="anchored" data-anchor-id="predictor-varies-at-level-1">Predictor varies at level 1</h4>
<p>Let’s start by imagining the following scenario: the <code>Treatment</code> variable is varying at our lowest level. Each patient receives only one type of treatment (A or B), but both treatment types are represented “within” each doctor and within each hospital:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/nested-patients2.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Scenario 1: predictor varies at level 1 (between patients, within doctors)</figcaption>
</figure>
</div>
<p>As a result, it would be inappropriate to ask <code>lme4</code> to fit random slopes for the <code>treatment</code> variable at the patient level. Instead, the “full” model (i.e., a model containing all of the possible fixed and random effects) would be the following:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>lme_health_slopes <span class="ot">&lt;-</span> <span class="fu">lmer</span>(outcome <span class="sc">~</span> treatment <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> treatment<span class="sc">|</span>doctor) <span class="sc">+</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                      (<span class="dv">1</span> <span class="sc">+</span> treatment<span class="sc">|</span>hospital), <span class="at">data =</span> health)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_health_slopes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: outcome ~ treatment + (1 + treatment | doctor) + (1 + treatment |  
    hospital)
   Data: health

REML criterion at convergence: 1834.1

Scaled residuals: 
     Min       1Q   Median       3Q      Max 
-2.38341 -0.60702 -0.02763  0.71956  2.88951 

Random effects:
 Groups   Name             Variance Std.Dev. Corr 
 doctor   (Intercept)       7.8846  2.8080        
          treatmentsurgery  7.3157  2.7047   -0.96
 hospital (Intercept)       0.5085  0.7131        
          treatmentsurgery  3.5595  1.8867   -0.91
 Residual                  23.5775  4.8557        
Number of obs: 300, groups:  doctor, 30; hospital, 5

Fixed effects:
                 Estimate Std. Error      df t value Pr(&gt;|t|)    
(Intercept)       26.6155     0.7223  4.0083  36.849 3.17e-06 ***
treatmentsurgery   6.2396     1.1270  3.9992   5.537  0.00521 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Correlation of Fixed Effects:
            (Intr)
trtmntsrgry -0.794</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This produces a model with four sets of random effects: two sets of random intercepts, and two sets of random slopes.</p>
</section>
<section id="predictor-varies-at-level-2" class="level4">
<h4 class="anchored" data-anchor-id="predictor-varies-at-level-2">Predictor varies at level 2</h4>
<p>Let’s now imagine a (perhaps more realistic) scenario. Each doctor is in fact a specialist in a certain type of treatment, but cannot deliver both. For this, we will need to read in the second version of our dataset.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>health2 <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/health2.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>If you look closely at the dataset, you can see that <code>treatment</code> does not vary within <code>doctor</code>; instead, it only varies within <code>hospital</code>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/nested-patients3.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Scenario 2: predictor varies at level 2 (between doctors, within hospitals)</figcaption>
</figure>
</div>
<p>This means we cannot fit random slopes for treatment at the second level any more. We have to drop our random slopes for <code>treatment</code> by <code>doctor</code>, like this:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>lme_health_slopes2 <span class="ot">&lt;-</span> <span class="fu">lmer</span>(outcome <span class="sc">~</span> treatment <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>doctor) <span class="sc">+</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>                      (<span class="dv">1</span> <span class="sc">+</span> treatment<span class="sc">|</span>hospital), <span class="at">data =</span> health2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>boundary (singular) fit: see help('isSingular')</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_health_slopes2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: outcome ~ treatment + (1 | doctor) + (1 + treatment | hospital)
   Data: health2

REML criterion at convergence: 1845.7

Scaled residuals: 
     Min       1Q   Median       3Q      Max 
-2.46342 -0.69900 -0.05043  0.69559  3.09601 

Random effects:
 Groups   Name             Variance Std.Dev. Corr 
 doctor   (Intercept)       6.318   2.514         
 hospital (Intercept)       1.372   1.171         
          treatmentsurgery  3.635   1.907    -1.00
 Residual                  24.417   4.941         
Number of obs: 300, groups:  doctor, 30; hospital, 5

Fixed effects:
                 Estimate Std. Error      df t value Pr(&gt;|t|)    
(Intercept)       25.7019     0.9265  4.3745  27.740 4.38e-06 ***
treatmentsurgery   4.5061     1.3766  4.0456   3.273   0.0302 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Correlation of Fixed Effects:
            (Intr)
trtmntsrgry -0.808
optimizer (nloptwrap) convergence code: 0 (OK)
boundary (singular) fit: see help('isSingular')</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This gives us three sets of random effects, as opposed to four.</p>
</section>
<section id="predictor-varies-at-level-3" class="level4">
<h4 class="anchored" data-anchor-id="predictor-varies-at-level-3">Predictor varies at level 3</h4>
<p>Finally, let’s imagine a scenario where each hospital is only equipped to offer one type of treatment (hopefully, not a realistic scenario!). Here, all doctors and patients within each hospital use exactly the same method.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/nested-patients4.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Scenario 3: predictor varies at level 3 (between hospitals)</figcaption>
</figure>
</div>
<p>At this stage, we can no longer include random slopes for the treatment predictor anywhere in our model. Each <code>hospital</code>, <code>doctor</code> and <code>patient</code> only experiences one of the two treatments, not both, so we have no variation between the treatments to estimate at any of these levels.</p>
<p>So, here, we would go back to our random intercepts only model.</p>
</section>
</section>
</section>
<section id="implicit-vs-explicit-nesting" class="level2" data-number="8.3">
<h2 data-number="8.3" class="anchored" data-anchor-id="implicit-vs-explicit-nesting"><span class="header-section-number">8.3</span> Implicit vs explicit nesting</h2>
<p>The different <code>health</code> datasets that have been explored above all have an important thing in common: the variables have been <strong>implicitly nested</strong>. Each new hospital, doctor and patient is given a unique identifier, to make it clear that doctors do not reoccur between hospitals, and patients do not reoccur between doctors.</p>
<p>In other words, all the information about the nesting is captured implicitly in the way that the data are coded.</p>
<p>However, you might sometimes be working with a dataset that has not been coded this way. So how do you deal with those situations? You have a few options:</p>
<ul>
<li>Recode your dataset so it is implicitly nested</li>
<li>Use explicit nesting in your <code>lme4</code> model formula</li>
<li>Use the <code>A/B</code> syntax in your <code>lme4</code> model formula</li>
</ul>
<section id="the-pastes-dataset" class="level3" data-number="8.3.1">
<h3 data-number="8.3.1" class="anchored" data-anchor-id="the-pastes-dataset"><span class="header-section-number">8.3.1</span> The Pastes dataset</h3>
<p>We’ll use another internal <code>lme4</code> dataset, the <code>Pastes</code> dataset, to show you what these three options look like in practice.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"Pastes"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Pastes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  strength batch cask sample
1     62.8     A    a    A:a
2     62.6     A    a    A:a
3     60.1     A    b    A:b
4     62.3     A    b    A:b
5     62.7     A    c    A:c
6     63.1     A    c    A:c</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This dataset is about measuring the strength of a chemical paste product, which is delivered in batches, each batch consisting of several casks. From ten random deliveries of the product, three casks were chosen at random (for a total of 30 casks). A sample was taken from each cask; from each sample, there were two assays, for a total of 60 assays.</p>
<p>There are four variables:</p>
<ul>
<li><code>strength</code>, paste strength, a continuous response variable; measured for each assay</li>
<li><code>batch</code>, delivery batch from which the sample was chosen (10 groups, A to J)</li>
<li><code>cask</code>, cask within the deliver batch from which the sample was chosen (3 groups, a to c)</li>
<li><code>sample</code>, batch &amp; cask combination (30 groups, A:a to J:c)</li>
</ul>
<p>The experimental design, when drawn out, looks like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/pastes_design.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Pastes dataset design</figcaption>
</figure>
</div>
<p>At first glance, this might look as if it’s a four-level model: assays within samples within casks within deliveries. However, that’s a bit of a overcomplication. There is only one <code>sample</code> collected per <code>cask</code>, meaning that we can really just think about assays being nested within casks directly.</p>
<p>The aim of this experiment was simply to assess the average strength of the chemical paste, adjusting for variation across batches and casks. Therefore, there are no fixed effects in the model - instead, we simply write <code>1</code> for the fixed portion of our model. We also won’t have any random slopes, only random intercepts.</p>
<p>If we follow the same procedure we did above for the <code>health</code> example, we might try something like this:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-7-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-1" role="tab" aria-controls="tabset-7-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-7-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-7-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>lme_paste <span class="ot">&lt;-</span> <span class="fu">lmer</span>(strength <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>cask), <span class="at">data =</span> Pastes)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_paste)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: strength ~ 1 + (1 | batch) + (1 | cask)
   Data: Pastes

REML criterion at convergence: 301.5

Scaled residuals: 
     Min       1Q   Median       3Q      Max 
-1.49025 -0.90096 -0.01247  0.62911  1.82246 

Random effects:
 Groups   Name        Variance Std.Dev.
 batch    (Intercept) 3.3639   1.8341  
 cask     (Intercept) 0.1487   0.3856  
 Residual             7.3060   2.7030  
Number of obs: 60, groups:  batch, 10; cask, 3

Fixed effects:
            Estimate Std. Error      df t value Pr(&gt;|t|)    
(Intercept)  60.0533     0.7125  6.7290   84.28 1.99e-11 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Something is wrong with this model. To spot it, we have to look carefully at the bottom of the random effects section, where it says <code>Number of obs</code> (short for observations).</p>
<p><code>lme4</code> has correctly identified that there are 10 delivery batches, and has fitted a set of 10 random intercepts for those batches - all good so far. However, R believes that we only have 3 casks, because the <code>cask</code> variable is implicitly nested, and so has only fitted a set of 3 random intercepts for that variable.</p>
<p>But this isn’t what we want. There is no link between cask A in batch A, and cask A in batch D - they have no reason to be more similar to each other than they are to other casks. We actually have 30 unique casks, and would like for each of them to have its own random intercept.</p>
</section>
<section id="recoding-for-implicit-nesting" class="level3" data-number="8.3.2">
<h3 data-number="8.3.2" class="anchored" data-anchor-id="recoding-for-implicit-nesting"><span class="header-section-number">8.3.2</span> Recoding for implicit nesting</h3>
<p>As shown above, the formula <code>strength ~ 1 + (1|batch) + (1|cask)</code> does not produce the model we want, because we don’t have implicit coding in the <code>cask</code> variable.</p>
<p>So, let’s create a new variable that gives unique values to each of the casks in our dataset. We’ll do this using the <code>mutate</code> function and the <code>paste()</code> function to create a unique ID for each cask.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-8-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-1" role="tab" aria-controls="tabset-8-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-8-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-8-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>Pastes <span class="ot">&lt;-</span> Pastes <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">unique_cask =</span> <span class="fu">paste</span>(batch, cask, <span class="at">sep =</span> <span class="st">"_"</span>))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Pastes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  strength batch cask sample unique_cask
1     62.8     A    a    A:a         A_a
2     62.6     A    a    A:a         A_a
3     60.1     A    b    A:b         A_b
4     62.3     A    b    A:b         A_b
5     62.7     A    c    A:c         A_c
6     63.1     A    c    A:c         A_c</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This generates 30 unique IDs, one for each of our unique casks. (We then have two observations of <code>strength</code> for each <code>unique_cask</code>.)</p>
<p>Now, we can go ahead and fit our desired model:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-9-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-1" role="tab" aria-controls="tabset-9-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-9-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-9-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>lme_paste_implicit <span class="ot">&lt;-</span> <span class="fu">lmer</span>(strength <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>unique_cask),</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">data =</span> Pastes)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_paste_implicit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: strength ~ 1 + (1 | batch) + (1 | unique_cask)
   Data: Pastes

REML criterion at convergence: 247

Scaled residuals: 
    Min      1Q  Median      3Q     Max 
-1.4798 -0.5156  0.0095  0.4720  1.3897 

Random effects:
 Groups      Name        Variance Std.Dev.
 unique_cask (Intercept) 8.434    2.9041  
 batch       (Intercept) 1.657    1.2874  
 Residual                0.678    0.8234  
Number of obs: 60, groups:  unique_cask, 30; batch, 10

Fixed effects:
            Estimate Std. Error      df t value Pr(&gt;|t|)    
(Intercept)  60.0533     0.6769  9.0000   88.72 1.49e-14 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>No error message this time, and it has correctly identified that there are 30 unique casks, from 10 different batches. We’ve solved the problem!</p>
<p>Incidentally, and which you may have already noticed, the recoding that we did above also perfectly replicates the existing <code>sample</code> variable. This means we would get an identical result if we fitted the model <code>strength ~ 1 + (1|batch) + (1|sample)</code> instead.</p>
</section>
<section id="fitting-a-model-with-explicit-nesting" class="level3" data-number="8.3.3">
<h3 data-number="8.3.3" class="anchored" data-anchor-id="fitting-a-model-with-explicit-nesting"><span class="header-section-number">8.3.3</span> Fitting a model with explicit nesting</h3>
<p>If we’re in a situation like the above where we don’t have nice, neat implicitly coded variables, but we don’t really want to spend loads of time recoding a bunch of variables, we can instead fit our model using explicit nesting in <code>lme4</code>.</p>
<p>That essentially means combining the recoding and model fitting steps, so that you don’t have to save a new variable.</p>
<p>For the <code>Pastes</code> dataset, it would look like this:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-10-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-1" role="tab" aria-controls="tabset-10-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-10-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-10-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>lme_paste_explicit <span class="ot">&lt;-</span> <span class="fu">lmer</span>(strength <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch<span class="sc">:</span>cask), <span class="at">data =</span> Pastes)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_paste_explicit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: strength ~ 1 + (1 | batch) + (1 | batch:cask)
   Data: Pastes

REML criterion at convergence: 247

Scaled residuals: 
    Min      1Q  Median      3Q     Max 
-1.4798 -0.5156  0.0095  0.4720  1.3897 

Random effects:
 Groups     Name        Variance Std.Dev.
 batch:cask (Intercept) 8.434    2.9041  
 batch      (Intercept) 1.657    1.2874  
 Residual               0.678    0.8234  
Number of obs: 60, groups:  batch:cask, 30; batch, 10

Fixed effects:
            Estimate Std. Error      df t value Pr(&gt;|t|)    
(Intercept)  60.0533     0.6769  9.0000   88.72 1.49e-14 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Notice that the output for this model and for the one right above are the same - it’s because they are exactly equivalent to each other! We used <code>batch:cask</code> to create <code>unique_cask</code> earlier, so we’ve just directly inserted that code into our formula.</p>
</section>
<section id="an-alternative-approach-the-ab-syntax" class="level3" data-number="8.3.4">
<h3 data-number="8.3.4" class="anchored" data-anchor-id="an-alternative-approach-the-ab-syntax"><span class="header-section-number">8.3.4</span> An alternative approach: the A/B syntax</h3>
<p>There is another way to deal with nested random effects that haven’t been implicitly coded into the dataset. It’s not necessarily the way that we would recommend - recoding your variables and/or using explicit nesting has far less potential to trip you up and go wrong - but we’ll introduce it briefly here, since it’s something you’re likely to see if you start working with mixed models a lot.</p>
<p>We could fit the same model to the <code>Pastes</code> dataset, and achieve a set of 30 intercepts for <code>cask</code> and 10 for <code>batch</code>, without making use of the <code>sample</code> variable or using the <code>A:B</code> notation.</p>
<p>It works like this: when random effect B is nested inside random effect A, you can simply write <code>A/B</code> on the right hand side of the <code>|</code>.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-11-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-1" role="tab" aria-controls="tabset-11-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-11-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-11-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>lme_paste_shorthand <span class="ot">&lt;-</span> <span class="fu">lmer</span>(strength <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch<span class="sc">/</span>cask), <span class="at">data =</span> Pastes)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_paste_shorthand)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: strength ~ 1 + (1 | batch/cask)
   Data: Pastes

REML criterion at convergence: 247

Scaled residuals: 
    Min      1Q  Median      3Q     Max 
-1.4798 -0.5156  0.0095  0.4720  1.3897 

Random effects:
 Groups     Name        Variance Std.Dev.
 cask:batch (Intercept) 8.434    2.9041  
 batch      (Intercept) 1.657    1.2874  
 Residual               0.678    0.8234  
Number of obs: 60, groups:  cask:batch, 30; batch, 10

Fixed effects:
            Estimate Std. Error      df t value Pr(&gt;|t|)    
(Intercept)  60.0533     0.6769  9.0000   88.72 1.49e-14 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This gives you an identical model output to both of the other two models we’ve tried. That’s because <code>(1|batch/cask)</code> is actually shorthand for <code>(1|batch) + (1|batch:cask)</code>, and in this dataset as we’ve seen above, <code>batch:cask</code> is the same as <code>sample</code> and <code>unique_cask</code>. In other words, we’ve fitted exactly the same model each time.</p>
</section>
<section id="which-option-is-best" class="level3" data-number="8.3.5">
<h3 data-number="8.3.5" class="anchored" data-anchor-id="which-option-is-best"><span class="header-section-number">8.3.5</span> Which option is best?</h3>
<div class="callout callout-style-simple callout-note no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<p>The <code>A/B</code> syntax is popular, and you will see it used, because it’s quick to write. But there are a couple of reasons that we would recommend you steer away from it:</p>
<ul>
<li><p><strong>It has more potential to go wrong.</strong> <code>A/B</code> is order-dependent, meaning that <code>A/B</code> <span class="math inline">\(\neq\)</span> <code>B/A</code>, and if you get that wrong, your model won’t work as intended. In contrast, explicit coding is order-invariant (<code>A:B + A</code> = <code>A + B:A</code>). Likewise, if you’ve implicitly coded your dataset, you can write your random effects in any order.</p></li>
<li><p><strong>It’s harder to interpret.</strong> Using implicit or explicit nesting gives you a separate <code>(1 + x|y)</code> structure in the formula for each clustering variable, which is more transparent when figuring out how many random effects you’ve fitted. Separate structures can also be more flexible, as you’ll see in at least one example later in the course.</p></li>
</ul>
</div>
</div>
</div>
<p>Overall: implicit coding of your dataset is best, and it’s best to do this implicit coding during data collection itself. It prevents mistakes, because the experimental design is clear from the data structure. It’s also the easiest and most flexible in terms of coding in <code>lme4</code>.</p>
<p>In the absence of an implicitly coded dataset, we strongly recommend sticking with explicit coding rather than the shorthand syntax - yes, it requires more typing, but it’s somewhat safer.</p>
<p>And, no matter which method you choose, always check the model output to see that the number of groups per clustering variables matches what you expect to see.</p>
</section>
</section>
<section id="exercises" class="level2" data-number="8.4">
<h2 data-number="8.4" class="anchored" data-anchor-id="exercises"><span class="header-section-number">8.4</span> Exercises</h2>
<section id="sec-exr_cake" class="level3" data-number="8.4.1">
<h3 data-number="8.4.1" class="anchored" data-anchor-id="sec-exr_cake"><span class="header-section-number">8.4.1</span> Cake</h3>
<div class="callout callout-style-default callout-exercise callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p><font style="font-variant: small-caps">Level:</font> <i class="fa-solid fa-star"></i><i class="fa-solid fa-star"></i><i class="fa-regular fa-star"></i><br></p>
<p>For this exercise, we’ll use the most delicious of the internal <code>lme4</code> datasets: <code>cake</code>.</p>
<p>This is a real dataset, taken from a thesis by Cook (<a href="https://www.sumsar.net/blog/source-of-the-cake-dataset/cook_1938_chocolate_cake.pdf">1983</a>), and is all about measuring cake quality of chocolate cakes baked using different recipes at particular temperatures.</p>
<p>Cook’s experiment worked as follows: for each recipe, she prepared 15 batches of cake mixture. Each batch was divided into 6 cakes, and each of those cakes were baked at one of the 6 temperatures.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-12-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-1" role="tab" aria-controls="tabset-12-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-12-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-12-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"cake"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>There are five variables in this dataset:</p>
<ul>
<li><code>recipe</code>, the exact recipe used for the cake (3 categories)</li>
<li><code>replicate</code>, the batter batch number for each recipe (15 replicates per recipe)</li>
<li><code>temperature</code>, a factor indicating which of 6 temperatures the cake was baked at</li>
<li><code>temp</code>, the numeric value of the baking temperature</li>
<li><code>angle</code>, the angle at which the cake broke (used as a measure of cake “tenderness”)</li>
</ul>
<p>For this exercise:</p>
<ol type="1">
<li>Sketch a graphic/diagram that captures the experimental design</li>
<li>Figure out what level of the dataset your variables of interest are varying at</li>
<li>Consider how you might recode the dataset to reflect implicit nesting</li>
<li>Fit and test at least one appropriate model</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Worked answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="consider-the-experimental-design" class="level4">
<h4 class="anchored" data-anchor-id="consider-the-experimental-design">Consider the experimental design</h4>
<p>The first thing to do is to draw out a diagram that captures the experimental design. That might look something like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/cake_design.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Experimental design for Cook’s cakes</figcaption>
</figure>
</div>
<p>We’ve got a 3-level dataset here: individual cakes within batches (replicates) within recipes. There are two fixed effects of interest - <code>recipe</code> and <code>temperature</code>, both categorical.</p>
<p>The next thing to think about is whether the coding within the dataset accurately reflects what’s going on.</p>
<p>There are 3 recipes, and 15 replicates of each recipe are mixed, for a total of 45 unique batches or mixtures. In the dataset as we’ve got it, the numbering has been repeated, so the nesting is not implicitly coded; but of course, replicate 1 for recipe A doesn’t have anything more in common with replicate 1 for recipe C.</p>
</section>
<section id="recode-the-dataset" class="level4">
<h4 class="anchored" data-anchor-id="recode-the-dataset">Recode the dataset</h4>
<p>So, let’s code up a new variable that captures unique replicates, and we’ll call it <code>batch</code>:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-13-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-1" role="tab" aria-controls="tabset-13-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-13-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-13-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>cake <span class="ot">&lt;-</span> cake <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">batch =</span> recipe<span class="sc">:</span>replicate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>As you should be able to see in your global environment, the new <code>batch</code> variable is a factor with 45 levels.</p>
<p>Each batch is then split into individual cakes, which undergo one of the 6 <code>temperature</code> treatments, for a total of 270 measurements.</p>
</section>
<section id="fit-a-model" class="level4">
<h4 class="anchored" data-anchor-id="fit-a-model">Fit a model</h4>
<p>Now, we can try fitting a model. We know that we want <code>recipe</code> and <code>temperature</code> as fixed effects, and probably also their interaction, at least to start with. We know we want to treat <code>batch</code> as a random effect (<code>replicate</code> nested within <code>recipe</code>), so we’ll include random intercepts.</p>
<p>We don’t, however, want to treat <code>recipe</code> as a random effect itself. It only has three levels, so it wouldn’t work well even if we wanted to. Plus, we’re specifically interested in those three levels and the differences between them.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-14-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-14-1" role="tab" aria-controls="tabset-14-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-14-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-14-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>lme_cake <span class="ot">&lt;-</span> <span class="fu">lmer</span>(angle <span class="sc">~</span> recipe<span class="sc">*</span>temperature <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch), <span class="at">data =</span> cake)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(lme_cake)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML. t-tests use Satterthwaite's method [
lmerModLmerTest]
Formula: angle ~ recipe * temperature + (1 | batch)
   Data: cake

REML criterion at convergence: 1638.6

Scaled residuals: 
     Min       1Q   Median       3Q      Max 
-2.64661 -0.61082 -0.05207  0.56985  2.75374 

Random effects:
 Groups   Name        Variance Std.Dev.
 batch    (Intercept) 41.84    6.468   
 Residual             20.47    4.524   
Number of obs: 270, groups:  batch, 45

Fixed effects:
                       Estimate Std. Error        df t value Pr(&gt;|t|)    
(Intercept)            33.12222    1.73683  42.00000  19.070  &lt; 2e-16 ***
recipeB                -1.47778    2.45625  42.00000  -0.602  0.55065    
recipeC                -1.52222    2.45625  42.00000  -0.620  0.53878    
temperature.L           6.43033    1.16822 210.00000   5.504 1.07e-07 ***
temperature.Q          -0.71285    1.16822 210.00000  -0.610  0.54239    
temperature.C          -2.32551    1.16822 210.00000  -1.991  0.04782 *  
temperature^4          -3.35128    1.16822 210.00000  -2.869  0.00454 ** 
temperature^5          -0.15119    1.16822 210.00000  -0.129  0.89715    
recipeB:temperature.L   0.45419    1.65211 210.00000   0.275  0.78365    
recipeC:temperature.L   0.08765    1.65211 210.00000   0.053  0.95774    
recipeB:temperature.Q  -0.23277    1.65211 210.00000  -0.141  0.88809    
recipeC:temperature.Q   1.21475    1.65211 210.00000   0.735  0.46299    
recipeB:temperature.C   2.69322    1.65211 210.00000   1.630  0.10456    
recipeC:temperature.C   2.63856    1.65211 210.00000   1.597  0.11175    
recipeB:temperature^4   3.02372    1.65211 210.00000   1.830  0.06863 .  
recipeC:temperature^4   3.13711    1.65211 210.00000   1.899  0.05895 .  
recipeB:temperature^5  -0.66354    1.65211 210.00000  -0.402  0.68836    
recipeC:temperature^5  -1.62525    1.65211 210.00000  -0.984  0.32637    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Correlation matrix not shown by default, as p = 18 &gt; 12.
Use print(x, correlation=TRUE)  or
    vcov(x)        if you need it</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="alternative-models" class="level4">
<h4 class="anchored" data-anchor-id="alternative-models">Alternative models</h4>
<p>If you want to do a bit of significance testing, you can try a few other versions of the model with different structures:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-15-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-15-1" role="tab" aria-controls="tabset-15-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-15-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-15-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>lme_cake2 <span class="ot">&lt;-</span> <span class="fu">lmer</span>(angle <span class="sc">~</span> recipe <span class="sc">+</span> temperature <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch), <span class="at">data =</span> cake)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>lme_cake3 <span class="ot">&lt;-</span> <span class="fu">lmer</span>(angle <span class="sc">~</span> recipe <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch), <span class="at">data =</span> cake)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>lme_cake4 <span class="ot">&lt;-</span> <span class="fu">lmer</span>(angle <span class="sc">~</span> temperature <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>batch), <span class="at">data =</span> cake)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>lm_cake <span class="ot">&lt;-</span> <span class="fu">lm</span>(angle <span class="sc">~</span> recipe<span class="sc">*</span>temperature, <span class="at">data =</span> cake)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(lme_cake, lm_cake) <span class="co"># random effects dropped</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>refitting model(s) with ML (instead of REML)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Data: cake
Models:
lm_cake: angle ~ recipe * temperature
lme_cake: angle ~ recipe * temperature + (1 | batch)
         npar    AIC    BIC  logLik deviance  Chisq Df Pr(&gt;Chisq)    
lm_cake    19 1901.3 1969.6 -931.63   1863.3                         
lme_cake   20 1719.0 1791.0 -839.53   1679.0 184.21  1  &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(lme_cake, lme_cake2) <span class="co"># recipe:temperature dropped</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>refitting model(s) with ML (instead of REML)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Data: cake
Models:
lme_cake2: angle ~ recipe + temperature + (1 | batch)
lme_cake: angle ~ recipe * temperature + (1 | batch)
          npar    AIC    BIC  logLik deviance Chisq Df Pr(&gt;Chisq)
lme_cake2   10 1709.6 1745.6 -844.79   1689.6                    
lme_cake    20 1719.0 1791.0 -839.53   1679.0 10.53 10     0.3953</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(lme_cake, lme_cake3) <span class="co"># temperature &amp; recipe:temperature dropped</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>refitting model(s) with ML (instead of REML)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Data: cake
Models:
lme_cake3: angle ~ recipe + (1 | batch)
lme_cake: angle ~ recipe * temperature + (1 | batch)
          npar    AIC    BIC  logLik deviance  Chisq Df Pr(&gt;Chisq)    
lme_cake3    5 1785.7 1803.7 -887.84   1775.7                         
lme_cake    20 1719.0 1791.0 -839.53   1679.0 96.636 15  5.642e-14 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(lme_cake, lme_cake4) <span class="co"># recipe &amp; recipe:temperature dropped</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>refitting model(s) with ML (instead of REML)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Data: cake
Models:
lme_cake4: angle ~ temperature + (1 | batch)
lme_cake: angle ~ recipe * temperature + (1 | batch)
          npar    AIC    BIC  logLik deviance Chisq Df Pr(&gt;Chisq)
lme_cake4    8 1706.1 1734.9 -845.06   1690.1                    
lme_cake    20 1719.0 1791.0 -839.53   1679.0 11.06 12     0.5238</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We see that when we drop the random intercepts for <code>batch</code>, and when we drop the <code>temperature</code> predictor, our chi-square values are significant. This indicates that these predictors are important. But we can drop <code>recipe</code> and <code>recipe:temperature</code> without a particularly big change in deviance.</p>
<p>(This is borne out somewhat if you’ve used the <code>lmerTest</code> package to perform degrees of freedom approximation and extract p-values as part of the <code>lme_cake</code> model summary.)</p>
<p>So, our final model is probably <code>lme_cake4 = angle ~ temperature + (1|batch)</code>.</p>
</section>
<section id="check-assumptions" class="level4">
<h4 class="anchored" data-anchor-id="check-assumptions">Check assumptions</h4>
<p>For completeness, we’ll check the assumptions of <code>lme_cake4</code> and visualise it for the sake of aiding our interpretation.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-16-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-16-1" role="tab" aria-controls="tabset-16-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-16-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-16-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(lme_cake4, </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">check =</span> <span class="fu">c</span>(<span class="st">"linearity"</span>, <span class="st">"homogeneity"</span>, <span class="st">"qq"</span>, <span class="st">"outliers"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(lme_cake4, </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">check =</span> <span class="fu">c</span>(<span class="st">"reqq"</span>, <span class="st">"pp_check"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Most of the assumptions look okay, with the exception of the normal Q-Q plot for the random intercepts. The set of intercepts doesn’t really look like it’s nicely normally distributed here. Maybe a more complicated mixed effects model (something beyond the linear type we’re going with here) would help. Or, maybe this just means we should be a little less decisive in our overall conclusions.</p>
</section>
<section id="visualise-the-model" class="level4">
<h4 class="anchored" data-anchor-id="visualise-the-model">Visualise the model</h4>
<p>Last but not least, we can visualise the model:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-17-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-17-1" role="tab" aria-controls="tabset-17-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-17-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-17-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">augment</span>(lme_cake4), <span class="fu">aes</span>(<span class="at">x =</span> temperature, <span class="at">y =</span> angle)) <span class="sc">+</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> .fitted, <span class="at">group =</span> batch), <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Overall, <code>angle</code> increases with <code>temperature</code>. (From what I understand of reading the thesis, this is a good thing from the perspective of the cake quality, as it suggests the cake is more tender. Scientific <em>and</em> delicious.)</p>
<p>We can see visually that the <code>recipe</code> and <code>recipe:temperature</code> terms don’t have much explanatory power by visualising the full model (commenting out the <code>facet_wrap</code> may also help you to see this):</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-18-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-18-1" role="tab" aria-controls="tabset-18-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-18-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-18-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">augment</span>(lme_cake), <span class="fu">aes</span>(<span class="at">x =</span> temperature, <span class="at">y =</span> angle, <span class="at">colour =</span> recipe)) <span class="sc">+</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> recipe) <span class="sc">+</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> .fitted, <span class="at">group =</span> batch))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-exercise callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus questions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p><font style="font-variant: small-caps">Level:</font> <i class="fa-solid fa-star"></i><i class="fa-solid fa-star"></i><i class="fa-regular fa-star"></i><br></p>
<p>If you want to think a bit harder about this dataset, consider these additional questions. Feel free to chat about them with a neighbour or with a trainer.</p>
<ul>
<li>Why doesn’t it work if you try to fit random slopes for <code>temperature</code> on <code>batch</code>? Have a look at the warning message that R gives you in this situation.</li>
<li>What happens if you use the numerical <code>temp</code> variable instead of the categorical <code>temperature</code>? Does it change your conclusions? Why might you prefer to use the numerical/continuous version?</li>
<li>Could <code>temperature</code> be treated as a random effect, under certain interpretations of the original research question? Is it possible or sensible to do that with the current dataset?</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>For more information on the very best way to bake a chocolate cake (and a lovely demonstration at the end about the dangers of extrapolating from a linear model), <a href="https://www.sumsar.net/blog/source-of-the-cake-dataset/">this blog post</a> is a nice source. It’s written by a data scientist who was so curious about the quirky <code>cake</code> dataset that he contacted Iowa State University, who helped him unearth Cook’s original thesis.</p>
</section>
<section id="sec-exr_parallel" class="level3" data-number="8.4.2">
<h3 data-number="8.4.2" class="anchored" data-anchor-id="sec-exr_parallel"><span class="header-section-number">8.4.2</span> Parallel fibres</h3>
<div class="callout callout-style-default callout-exercise callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p><font style="font-variant: small-caps">Level:</font> <i class="fa-solid fa-star"></i><i class="fa-solid fa-star"></i><i class="fa-regular fa-star"></i><br></p>
<p>For this exercise, we’ll be using a neurohistology dataset that focuses on a particular type of neuron found in the cerebellum, known as a parallel fibre. Parallel fibres are found in the uppermost layer of cerebellar cortex, and are known for being long; this experiment was designed to test whether the depth at which the fibre was found, had a bearing on its length.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-19-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-19-1" role="tab" aria-controls="tabset-19-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-19-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-19-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>parallel <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/parallel.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>To measure the length of the fibres (which are &lt;10mm long - big for a neuron!), slices were taken from the cerebella of six cats, at different depths. The depth of each slice was recorded. These slices were then stained, so that individual parallel fibres could be identified and measured.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/cerebellum_histology.jpg" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption>An example of a stained slice from cerebellar cortex</figcaption>
</figure>
</div>
<p>The dataset contains five variables:</p>
<ul>
<li><code>length</code> of the fibre (in micrometres)</li>
<li><code>depth</code> of the slice (in micrometres)</li>
<li><code>fibre</code>, individual IDs for all of the fibres</li>
<li><code>slice</code> ID number (maximum 10 slices per cat)</li>
<li><code>cat</code> ID number (1 through 6)</li>
</ul>
<p>For this exercise:</p>
<ol type="1">
<li>Sketch a graphic/diagram that captures the experimental design</li>
<li>Determine whether the dataset requires recoding or explicit nesting</li>
<li>Fit and test at least one appropriate model</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Worked answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="visualise-the-design" class="level4">
<h4 class="anchored" data-anchor-id="visualise-the-design">Visualise the design</h4>
<p>This is a nested design with three levels: <code>fibre</code> within <code>slice</code> within <code>cat</code>. But the fixed predictor <code>depth</code> varies at level 2, between slices (not between fibres).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images_mixed-effects/neurohist_design.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Experimental design</figcaption>
</figure>
</div>
</section>
<section id="recoding" class="level4">
<h4 class="anchored" data-anchor-id="recoding">Recoding</h4>
<p>If we look at the structure of the dataset, we can see that the numbering for the <code>slice</code> variable starts again for each <code>cat</code> at 1, which is not what we want.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-20-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-20-1" role="tab" aria-controls="tabset-20-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-20-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-20-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>parallel <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">46</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 5
  fibre length depth slice   cat
  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1   5780   260     1     1
2     2   5730   260     1     1
3     3   5790   260     1     1
4     4   5860   260     1     1
5     5   5690   260     1     1
6     6   5940   260     1     1
7     7   5950   260     1     1
8     8   5940   290     2     1
9    46   4670   300     1     2</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>So, we need to recode a new variable, or be prepared to use explicit nesting in our model formula. Note that for this recoding to work, we also need to ask R to treat <code>slice</code> and <code>cat</code> as factors rather than numeric variables.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-21-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-21-1" role="tab" aria-controls="tabset-21-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-21-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-21-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>parallel <span class="ot">&lt;-</span> parallel <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cat =</span> <span class="fu">as.factor</span>(cat)) <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">slice =</span> <span class="fu">as.factor</span>(slice)) <span class="sc">%&gt;%</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">unique_slice =</span> slice<span class="sc">:</span>cat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="fit-a-model-1" class="level4">
<h4 class="anchored" data-anchor-id="fit-a-model-1">Fit a model</h4>
<p>The full model that we could fit to these data contains three random effects: random intercepts for <code>unique_slice</code> (or <code>slice:cat</code> if you’re explicitly coding), random intercepts for <code>cat</code>, and random slopes for <code>depth</code> on <code>cat</code>.</p>
<p>(Since <code>depth</code> doesn’t vary within <code>slice</code>, i.e., each <code>slice</code> has only one <code>depth</code>, we can’t fit random slopes at level 2.)</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-22-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-22-1" role="tab" aria-controls="tabset-22-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-22-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-22-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>lme_parallel <span class="ot">&lt;-</span> <span class="fu">lmer</span>(length <span class="sc">~</span> depth <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>slice<span class="sc">:</span>cat) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> depth<span class="sc">|</span>cat), </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">data =</span> parallel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv, :
Model failed to converge with max|grad| = 1.2981 (tol = 0.002, component 1)</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>You may notice that you get an error - the model fails to converge. There are a couple of fixes we could try that involve tweaking the settings in the estimation procedure (e.g., increasing the maximum number of iterations allowed).</p>
<p>However, most errors like this just mean that we’re being too ambitious. So, the approach we’ll take here is to make the model simpler.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-23-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-23-1" role="tab" aria-controls="tabset-23-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-23-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-23-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>lme_parallel_int <span class="ot">&lt;-</span> <span class="fu">lmer</span>(length <span class="sc">~</span> depth <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>slice<span class="sc">:</span>cat) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>cat), </span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">data =</span> parallel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="check-the-assumptions" class="level4">
<h4 class="anchored" data-anchor-id="check-the-assumptions">Check the assumptions</h4>
<p>Next, we check the assumptions of our intercepts-only nested model:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-24-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-24-1" role="tab" aria-controls="tabset-24-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-24-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-24-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(lme_parallel_int, </span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">check =</span> <span class="fu">c</span>(<span class="st">"linearity"</span>, <span class="st">"homogeneity"</span>, <span class="st">"qq"</span>, <span class="st">"outliers"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(lme_parallel_int, </span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">check =</span> <span class="fu">c</span>(<span class="st">"reqq"</span>, <span class="st">"pp_check"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-25-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Not bad at all. There are no obvious errors cropping up in these plots.</p>
</section>
<section id="visualise-the-model-1" class="level4">
<h4 class="anchored" data-anchor-id="visualise-the-model-1">Visualise the model</h4>
<p>Last but not least, we should have a look at our model predictions visually.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-25-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-25-1" role="tab" aria-controls="tabset-25-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-25-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-25-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">augment</span>(lme_parallel_int), <span class="fu">aes</span>(<span class="at">x =</span> depth, <span class="at">y =</span> length, <span class="at">colour =</span> cat)) <span class="sc">+</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> .fitted, <span class="at">group =</span> cat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Despite the fact that <code>depth</code> is a continuous variable, this plot still has jagged, rather than straight, lines of best fit. This is because the plot is also taking into account the multiple sets of random intercepts for <code>slice</code> that are contained within each <code>cat</code> cluster.</p>
<p>We can, however, extract just the set of intercepts by <code>cat</code>, and with a bit more fuss, use this to add lines to the plot with <code>geom_abline</code>:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-26-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-26-1" role="tab" aria-controls="tabset-26-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-26-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-26-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the coef function to extract the coefficients</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>cat_coefs <span class="ot">&lt;-</span> <span class="fu">coef</span>(lme_parallel_int)<span class="sc">$</span>cat</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="co"># use geom_abline to add individual lines for each cat</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">augment</span>(lme_parallel_int), <span class="fu">aes</span>(<span class="at">x =</span> depth, <span class="at">y =</span> length, <span class="at">colour =</span> cat)) <span class="sc">+</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> cat_coefs[<span class="dv">1</span>,<span class="dv">1</span>], <span class="at">slope =</span> cat_coefs[<span class="dv">1</span>,<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> cat_coefs[<span class="dv">2</span>,<span class="dv">1</span>], <span class="at">slope =</span> cat_coefs[<span class="dv">2</span>,<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> cat_coefs[<span class="dv">3</span>,<span class="dv">1</span>], <span class="at">slope =</span> cat_coefs[<span class="dv">3</span>,<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> cat_coefs[<span class="dv">4</span>,<span class="dv">1</span>], <span class="at">slope =</span> cat_coefs[<span class="dv">4</span>,<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> cat_coefs[<span class="dv">5</span>,<span class="dv">1</span>], <span class="at">slope =</span> cat_coefs[<span class="dv">5</span>,<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> cat_coefs[<span class="dv">6</span>,<span class="dv">1</span>], <span class="at">slope =</span> cat_coefs[<span class="dv">6</span>,<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="is-this-a-good-model" class="level4">
<h4 class="anchored" data-anchor-id="is-this-a-good-model">Is this a good model?</h4>
<div class="tabset-margin-container"></div><div class="panel-tabset" data-group="language">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-27-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-27-1" role="tab" aria-controls="tabset-27-1" aria-selected="true">R</a></li></ul>
<div class="tab-content" data-group="language">
<div id="tabset-27-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-27-1-tab">
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(parallel, <span class="fu">aes</span>(<span class="at">x =</span> depth, <span class="at">y =</span> length, <span class="at">colour =</span> slice)) <span class="sc">+</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>cat) <span class="sc">+</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="08-nested-random-effects_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>If we plot the data faceted by <code>cat</code>, it suggests that the relationship between <code>depth</code> and <code>length</code> varies between <code>cat</code>. But we were forced to drop the random slopes for <code>depth|cat</code> due to lack of model convergence.</p>
<p>Our diagnostic plots look pretty good for our simpler, intercepts-only model, but the raw data indicate we might be missing something. Do you still trust the model? If no, what might a researcher do to improve this analysis?</p>
</section>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-exercise callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus question: notation
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p><font style="font-variant: small-caps">Level:</font> <i class="fa-solid fa-star"></i><i class="fa-solid fa-star"></i><i class="fa-solid fa-star"></i><br></p>
<p>Think back to the brief introduction to linear mixed effects models notation given in section 5.5.1 of the course materials.</p>
<p>What would the equation of a three level model fitted to the <code>parallel</code> dataset look like?</p>
<p>Hint: you’ll need more subscript letters than you did for a two-level model!</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer: three-level intercepts-only
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>E.g., <code>length ~ depth + (1|slice:cat) + (1|cat)</code></p>
<p>Level 1:</p>
<p><span class="math display">\[
y_{ijk} = \beta_{0jk} + \beta_{1}x_{1ijk} + \epsilon_{ijk}
\]</span></p>
<p>Level 2:</p>
<p><span class="math display">\[
\beta_{0jk} = \delta_{00k} + U_{0jk}
\]</span></p>
<p>Level 3:</p>
<p><span class="math display">\[
\delta_{00k} = \gamma_{000} + V_{00k}
\]</span></p>
<p>where,</p>
<p><span class="math display">\[
\left( \begin{array}{c} U_{0jk} \end{array} \right) ∼ N \left( \begin{array}{c} 0 \end{array}   , \begin{array}{cc} \tau^2_{00} \end{array} \right)
\]</span></p>
<p>and,</p>
<p><span class="math display">\[
\left( \begin{array}{c} V_{00k} \end{array} \right) ∼ N \left( \begin{array}{c} 0 \end{array}   , \begin{array}{cc} \tau^2_{00} \end{array} \right)
\]</span></p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer: three-level intercepts &amp; slopes
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>E.g., <code>length ~ depth + (1|slice:cat) + (1 + depth|cat)</code></p>
<p>Level 1:</p>
<p><span class="math display">\[
y_{ijk} = \beta_{0jk} + \beta_{1k}x_{1ijk} + \epsilon_{ijk}
\]</span></p>
<p>Level 2:</p>
<p><span class="math display">\[
\beta_{0jk} = \beta_{00k} + U_{0jk}
\]</span></p>
<p>Level 3:</p>
<p><span class="math display">\[
\beta_{00k} = \gamma_{000} + V_{00k}
\]</span> <span class="math display">\[
\beta_{1k} = \gamma_{100} + V_{10k}
\]</span></p>
<p>Where,</p>
<p><span class="math display">\[
\left( \begin{array}{c} U_{0jk} \end{array} \right) ∼ N \left( \begin{array}{c} 0 \end{array}   , \begin{array}{cc} \tau^2_{00} \end{array} \right)
\]</span></p>
<p>and,</p>
<p><span class="math display">\[
\left( \begin{array}{c} V_{00k} \\ V_{10k} \end{array} \right) ∼ N \left( \begin{array}{c} 0 \\ 0 \end{array}   , \begin{array}{cc} \tau^2_{00} &amp; \rho_{01} \\ \rho_{01} &amp;  \tau^2_{10} \end{array} \right)
\]</span></p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="summary" class="level2" data-number="8.5">
<h2 data-number="8.5" class="anchored" data-anchor-id="summary"><span class="header-section-number">8.5</span> Summary</h2>
<p>Sometimes, a dataset contains multiple clustering variables. When one of those clustering variables is nested inside the other, we can model this effectively by estimating random effects at multiple levels.</p>
<p>Adding additional levels can create some complications, e.g., determining which level of the dataset your predictor variables are varying at. But it can also allow us to deal with real-life hierarchical data structures, which are common in research.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Key points
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Random effect B is nested inside random effect A, if each category of B occurs uniquely within only one category of A</li>
<li>It’s important to figure out what level of the hierarchy or model a predictor variable is varying at, to determine where random slopes are appropriate</li>
<li>Nested random effects can be implicitly or explicitly coded in a dataframe, which determines how the model should be specified in <code>lme4</code></li>
</ul>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../materials/07-checking-assumptions.html" class="pagination-link" aria-label="Checking assumptions">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Checking assumptions</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../materials/09-crossed-random-effects.html" class="pagination-link" aria-label="Crossed random effects">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Crossed random effects</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Bioinformatics Training Facility</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>